packages/dd-trace/src/opentelemetry/metrics/otlp_transformer.js

```diff
@@ -0,0 +1,255 @@
+'use strict'
+
+const OtlpTransformerBase = require('../otlp/otlp_transformer_base')
+const { getProtobufTypes } = require('../otlp/protobuf_loader')
+
+// Get the aggregation temporality enum
+const { protoAggregationTemporality } = getProtobufTypes()
+const AGGREGATION_TEMPORALITY_DELTA = protoAggregationTemporality.values.AGGREGATION_TEMPORALITY_DELTA
+
+/**
+ * OtlpTransformer transforms metrics to OTLP format.
+ *
+ * This implementation follows the OTLP Metrics v1.7.0 Data Model specification:
+ * https://opentelemetry.io/docs/specs/otlp/#metrics-data-model
+ *
+ * @class OtlpTransformer
+ * @extends OtlpTransformerBase
+ */
+class OtlpTransformer extends OtlpTransformerBase {
+  /**
+   * Creates a new OtlpTransformer instance.
+   *
+   * @param {import('@opentelemetry/api').Attributes} resourceAttributes - Resource attributes
+   * @param {string} protocol - OTLP protocol (http/protobuf or http/json)
+   */
+  constructor (resourceAttributes, protocol) {
+    super(resourceAttributes, protocol, 'metrics')
+  }
+
+  /**
+   * Transforms metrics to OTLP format based on the configured protocol.
+   * @param {Array} metrics - Array of metric data to transform
+   * @returns {Buffer} Transformed metrics in the appropriate format
+   */
+  transformMetrics (metrics) {
+    if (this.protocol === 'http/json') {
+      return this.#transformToJson(metrics)
+    }
+    return this.#transformToProtobuf(metrics)
+  }
+
+  /**
+   * Transforms metrics to protobuf format.
+   * @param {Array} metrics - Array of metrics to transform
+   * @returns {Buffer} Protobuf-encoded metrics
+   * @private
+   */
+  #transformToProtobuf (metrics) {
+    const { protoMetricsService } = getProtobufTypes()
+
+    const metricsData = {
+      resourceMetrics: [{
+        resource: this._transformResource(),
+        scopeMetrics: this.#transformScope(metrics),
+        schemaUrl: ''
+      }]
+    }
+
+    return this._serializeToProtobuf(protoMetricsService, metricsData)
+  }
+
+  /**
+   * Transforms metrics to JSON format.
+   * @param {Array} metrics - Array of metrics to transform
+   * @returns {Buffer} JSON-encoded metrics
+   * @private
+   */
+  #transformToJson (metrics) {
+    const metricsData = {
+      resourceMetrics: [{
+        resource: this._transformResource(),
+        scopeMetrics: this.#transformScope(metrics, true)
+      }]
+    }
+    return this._serializeToJson(metricsData)
+  }
+
+  /**
+   * Creates scope metrics grouped by instrumentation scope.
+   * @param {Array} metrics - Array of metrics to transform
+   * @param {boolean} isJson - Whether to format for JSON output
+   * @returns {Object[]} Array of scope metric objects
+   * @private
+   */
+  #transformScope (metrics, isJson = false) {
+    const groupedMetrics = this._groupByInstrumentationScope(metrics)
+    const scopeMetrics = []
+
+    for (const [key, metricsInScope] of groupedMetrics) {
+      const [name, version, schemaUrl] = key.split('@')
```

Comment:
Opentelemetry doesn't allow the `@` character in meter names, metric names, or Attributes. So it's a useful delimiter here